autotest/gcore/interpolateatpoint.py

```diff
@@ -0,0 +1,143 @@
+
+    res = ds.GetRasterBand(1).InterpolateAtPoint(1000, 120, gdal.GRIORA_Bilinear)
+    assert res == 0.0  # I would expect None. To be studied
+
+
+def test_interpolateatpoint_2_bands():
+
+    mem_ds = gdal.GetDriverByName("MEM").Create(
+        "", xsize=2, ysize=2, bands=2, eType=gdal.GDT_Float32
+    )
+    np = pytest.importorskip("numpy")
+    # First band with values values
+    raster_array_1 = np.array(([10.5, 1.3], [2.4, 3.8]))
+    mem_ds.GetRasterBand(1).WriteArray(raster_array_1)
+    # Second band with float values, also negative
+    raster_array_2 = np.array(([10.5, 1.3], [-2.4, 3.8]))
+    mem_ds.GetRasterBand(2).WriteArray(raster_array_2)
+
+    got_bilinear = mem_ds.GetRasterBand(1).InterpolateAtPoint(
+        1, 1, gdal.GRIORA_Bilinear
+    )
+    expected = np.sum(raster_array_1) / 4.0
+    assert got_bilinear == pytest.approx(expected, 1e-6)
+
+    got_bilinear = mem_ds.GetRasterBand(2).InterpolateAtPoint(
+        1, 1, gdal.GRIORA_Bilinear
+    )
+    expected = np.sum(raster_array_2) / 4.0
+    assert got_bilinear == pytest.approx(expected, 1e-6)
+
+
+def test_interpolateatpoint_bilinear_several_points():
+
+    mem_ds = gdal.GetDriverByName("MEM").Create(
+        "", xsize=3, ysize=2, bands=1, eType=gdal.GDT_Float32
+    )
+    np = pytest.importorskip("numpy")
+    raster_array_1 = np.array(([10.5, 1.3, 0.5], [2.4, 3.8, -1.0]))
+    mem_ds.GetRasterBand(1).WriteArray(raster_array_1)
+
+    got_bilinear = mem_ds.GetRasterBand(1).InterpolateAtPoint(
+        0.5, 0.5, gdal.GRIORA_Bilinear
+    )
+    assert got_bilinear == pytest.approx(10.5, 1e-6)
+
+    got_bilinear = mem_ds.GetRasterBand(1).InterpolateAtPoint(
+        1.5, 1.5, gdal.GRIORA_Bilinear
+    )
+    assert got_bilinear == pytest.approx(3.8, 1e-6)
+
+    got_bilinear = mem_ds.GetRasterBand(1).InterpolateAtPoint(
+        1.5, 1.5, gdal.GRIORA_Bilinear
+    )
+    assert got_bilinear == pytest.approx(3.8, 1e-6)
+
+    got_bilinear = mem_ds.GetRasterBand(1).InterpolateAtPoint(
+        2, 1, gdal.GRIORA_Bilinear
+    )
+    assert got_bilinear == pytest.approx((1.3 + 0.5 + 3.8 - 1) / 4, 1e-6)
+
+    # compute the linear intepolation manually
+    got_bilinear = mem_ds.GetRasterBand(1).InterpolateAtPoint(
+        2.1, 1.2, gdal.GRIORA_Bilinear
+    )
+    expected = 1.3 * 0.4 * 0.3 + 0.5 * 0.6 * 0.3 + 3.8 * 0.4 * 0.7 - 1 * 0.6 * 0.7
+    assert got_bilinear == pytest.approx(expected, 1e-6)
+
+
+def test_interpolateatpoint_cubic_several_points():
+
+    mem_ds = gdal.GetDriverByName("MEM").Create(
+        "", xsize=4, ysize=4, bands=1, eType=gdal.GDT_Float32
+    )
+    np = pytest.importorskip("numpy")
+    raster_array_1 = np.array(
+        (
+            [1.0, 2.0, 1.5, -0.3],
+            [1.0, 2.0, 1.5, -0.3],
+            [1.0, 2.0, 1.5, -0.3],
+            [1.0, 2.0, 1.5, -0.3],
+        )
+    )
+    mem_ds.GetRasterBand(1).WriteArray(raster_array_1)
+
+    # using https://tools.timodenk.com/cubic-spline-interpolation to get some samples
```

Comment:
2 potential reasons:
- perhaps because the cubic spline in this page is not the same method as the cubic of GDAL. The Cubic in GDAL is a Catmull-Rohm cubic (https://github.com/OSGeo/gdal/blob/b6b5546456fae7fff72560128d658c5e5ce597e5/alg/gdalwarpkernel.cpp#L3391)
- or the 0.5 pixel shift applied (cf comments "Convert from upper left corner of pixel coordinates to center of" in GDALRPCGetDEMHeight()). The reasoning is that the pixel value is exactly reached at the center of the pixel.

Reply:
I plotted the values with the three interpolations and got this:
(5 pixels in the x axis, the last two equal)
in any case, I will try to fix it in a later PR, because that interpolation affects RPC, and I do not want to add noise there.

![Figure_1](https://github.com/user-attachments/assets/66b9f6f3-23c4-4087-81e0-8e1375e6820d)


```python
def test_plot():
    import matplotlib.pyplot as plt

    mem_ds = gdal.GetDriverByName("MEM").Create(
        "", xsize=5, ysize=4, bands=1, eType=gdal.GDT_Float32
    )
    np = pytest.importorskip("numpy")
    raster_array_1 = np.array(
        (
            [1.0, 2.0, 1.5, 0.3, 0.3],
            [1.0, 2.0, 1.5, 0.3, 0.3],
            [1.0, 2.0, 1.5, 0.3, 0.3],
            [1.0, 2.0, 1.5, 0.3, 0.3],
        )
    )
    band = mem_ds.GetRasterBand(1)
    band.WriteArray(raster_array_1)

    t = np.arange(0.0, 5.0, 0.01)
    y = 1.5
    y_cubic = [band.InterpolateAtPoint(x, y, gdal.GRIORA_Cubic) for x in t]
    y_linear = [band.InterpolateAtPoint(x, y, gdal.GRIORA_Bilinear) for x in t]
    y_near = [band.InterpolateAtPoint(x, y, gdal.GRIORA_NearestNeighbour) for x in t]

    plt.plot(t, y_linear, "r", t, y_cubic, "b", t, y_near, "g")
    plt.grid(True)
    plt.show()
```

Reply:
> * The Cubic in GDAL is a Catmull-Rohm cubic

The code in `BiCubicKernel` seems to be similar to `GWKBSpline`
https://github.com/OSGeo/gdal/blob/b6b5546456fae7fff72560128d658c5e5ce597e5/alg/gdalwarpkernel.cpp#L3464

Reply:
> The code in `BiCubicKernel` seems to be similar to `GWKBSpline`

oh! you're right. I didn't realize that. This was introduced in 8a7c8dcd701 . Ok, so we should rather map GRIORA_CubicSpline to it, and rename it internally to CubicBSpline
It would be nice to add proper mapping for Catmull-Rohm cubic / GRIORA_Cubic, but that can be done later

Reply:
> > The code in `BiCubicKernel` seems to be similar to `GWKBSpline`
> 
> oh! you're right. I didn't realize that. This was introduced in [8a7c8dc](https://github.com/OSGeo/gdal/commit/8a7c8dcd7016cd1798b86001e427360bb5338363) . Ok, so we should rather map GRIORA_CubicSpline to it, and rename it internally to CubicBSpline It would be nice to add proper mapping for Catmull-Rohm cubic / GRIORA_Cubic, but that can be done later

That was my impression yesterday night. But it opens questions respect to RPC code.
I see three enums about interpolation methods (are there more?):

- `DEMResampleAlg`
- `GDALRIOResampleAlg`
- `GDALResampleAlg`

The last two have to be equivalent if I understand correctly the comment before their definitions.
The first one is used only in RPC, right?
But based on what we have seen here, `DEMResampleAlg::DRA_Cubic` is equivalent to `GDALResampleAlg::GRA_CubicSpline`

At `gdal_rpc.cpp` is the only place I found that something is set to `DRA_Cubic`:
```C++
    const char *pszDEMInterpolation =
        CSLFetchNameValueDef(papszOptions, "RPC_DEMINTERPOLATION", "bilinear");
    if (EQUAL(pszDEMInterpolation, "near"))
    {
        psTransform->eResampleAlg = DRA_NearestNeighbour;
    }
    else if (EQUAL(pszDEMInterpolation, "bilinear"))
    {
        psTransform->eResampleAlg = DRA_Bilinear;
    }
    else if (EQUAL(pszDEMInterpolation, "cubic"))
    {
        psTransform->eResampleAlg = DRA_Cubic;
    }
```
that is "conflicting" with other places (where we use the `-r` argument) about the meaning of `cubic`.

So, me guess about what to do is:

- Implement an interpolation in `InterpolateAtPoint` for `GDALResampleAlg::GRA_Cubic` for `cubic` and rename the current one for `cubicspline`.
- When RPC is using `DRA_Cubic`, use `GRA_CubicSpline`. That will keep the current (3.9) behavior. Would it be a good idea to change that in GDAL 4.0, so RPC would have `cubic` and `cubicspline`, consistent with other places in the code?
- Change the API of the internal function `GDALInterpolateAtPoint` to use `GDALRIOResampleAlg` or  `GDALResampleAlg` instead of `DEMResampleAlg`...
- Question: What is the difference between `GDALRIOResampleAlg` and `GDALResampleAlg`? Which one should I use in the internal function `GDALInterpolateAtPoint`?

Reply:
> The last two have to be equivalent if I understand correctly the comment before their definitions.

The values of the constants are designed to match when they are available in both, but some methods are present in one and not in the other one

> The first one is used only in RPC, right?

yes was purely an implementation detail of the RPC code

> But based on what we have seen here, DEMResampleAlg::DRA_Cubic is equivalent to GDALResampleAlg::GRA_CubicSpline

yes apparently.  so DRA_Cubic should be probably renamed to DRA_CubicSpline for consistence

> that is "conflicting" with other places (where we use the -r argument) about the meaning of cubic.

indeed. For existing RPC_DEMINTERPOLATION=cubic, we can't change the name. But for the new -r flag of gdallocationinfo, we should accept "cubicspline" for now, and reserve "cubic" for Catmull-Rohm cubic if we implement it at some later time

> Change the API of the internal function GDALInterpolateAtPoint to use GDALRIOResampleAlg

yes we could do that to avoid adding new enumerations.

> What is the difference between GDALRIOResampleAlg and GDALResampleAlg?

GDALRIOResampleAlg is for the methods supported by RasterIO() and the overview building code
GDALResampleAlg is for the warping code

> Which one should I use in the internal function GDALInterpolateAtPoint?

I would go for GDALRIOResampleAlg